In [1]:
# doing the minimum necessary imports
# more modules would be imported as and when needed

import pandas as pd  
import numpy as np  
import matplotlib.pyplot as plt  
%matplotlib inline

# reading data from CSV file. 
# reading bank currency note data into pandas dataframe.
bankdata = pd.read_csv("bill_authentication.csv")  

# Exploratory Data Analysis
print(bankdata.shape)  
print("------------")

#bankdata.head()

# shuffling the 100% of the data
print(bankdata.sample(random_state=100, frac=1).head(10)) 

## shuffle the original dataframe
# bankdata = bankdata.sample(random_state=100, frac=1)

(1372, 5)
------------
      Variance  Skewness  Curtosis  Entropy  Class
1058  -1.56210   -2.2121   4.25910  0.27972      1
714    2.55590    3.3605   2.03210  0.26809      0
1061  -2.31470    3.6668  -0.69690 -1.24740      1
399    2.96950    5.6222   0.27561 -1.15560      0
382    0.86202    2.6963   4.29080  0.54739      0
376    3.23030    7.8384  -3.53480 -1.21510      0
987   -0.55648    3.2136  -3.30850 -2.79650      1
416    4.34830   11.1079  -4.08570 -4.25390      0
945   -1.76970    3.4329  -1.21440 -2.37890      1
595    3.18360    7.2321  -1.07130 -2.59090      0


In [2]:
bankdata.info()  # this helps in finding any missing values

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1372 entries, 0 to 1371
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Variance  1372 non-null   float64
 1   Skewness  1372 non-null   float64
 2   Curtosis  1372 non-null   float64
 3   Entropy   1372 non-null   float64
 4   Class     1372 non-null   int64  
dtypes: float64(4), int64(1)
memory usage: 53.7 KB


<b> Analysis : </b> Their is no missing data. This data is clean.

In [3]:
# Data Preprocessing
# Data preprocessing involves 
# (1) Dividing the data into attributes and labels and 
# (2) dividing the data into training and testing sets.

# To divide the data into attributes and labels, do :
X = bankdata.drop('Class', axis=1)  
y = bankdata['Class']  

# the final preprocessing step is to divide data into training and test sets
from sklearn.model_selection import train_test_split  
X_train, X_test, y_train, y_test = train_test_split(X,y,random_state=100)
# default test_size parameter value is 0.25

# Training the Algorithm. Here we would use DecisionTreeClassifier
from sklearn.tree import DecisionTreeClassifier  
classifier = DecisionTreeClassifier()  
classifier.fit(X_train, y_train)

# make predictions on the test data
y_pred = classifier.predict(X_test)

# Evaluating the Algorithm
from sklearn.metrics import classification_report, confusion_matrix, accuracy_score  
print(confusion_matrix(y_test, y_pred))  
print(classification_report(y_test, y_pred))
print(accuracy_score(y_test, y_pred))

# Remember : for evaluating classification-based ML algo use  
# confusion_matrix, classification_report and accuracy_score.

# And for evaluating regression-based ML Algo use Mean Squared Error(MSE) 
# or RMSE (Root Mean Squared Error), ...

[[197   1]
 [  3 142]]
              precision    recall  f1-score   support

           0       0.98      0.99      0.99       198
           1       0.99      0.98      0.99       145

    accuracy                           0.99       343
   macro avg       0.99      0.99      0.99       343
weighted avg       0.99      0.99      0.99       343

0.9883381924198251


<b><font color='green'>Analysis</font></b> : From the confusion matrix, you can see that out of 343 test instances, our algorithm misclassified only 4. This is approx 99% accuracy. 

In [4]:
# Importing Libraries
import pandas as pd  
import numpy as np  
import matplotlib.pyplot as plt  
%matplotlib inline

# Importing the Dataset
dataset = pd.read_csv('petrol_consumption.csv')

dataset.head()  

Petrol_tax  Average_income  Paved_Highways  Population_Driver_licence(%)  \
0         9.0            3571            1976                         0.525   
1         9.0            4092            1250                         0.572   
2         9.0            3865            1586                         0.580   
3         7.5            4870            2351                         0.529   
4         8.0            4399             431                         0.544   

   Petrol_Consumption  
0                 541  
1                 524  
2                 561  
3                 414  
4                 410

In [5]:
# To see statistical details of the dataset, execute the following command:

#dataset.describe()
dataset['Petrol_Consumption'].mean()  # avg of the target var. 

576.7708333333334

In [6]:
# Preparing the Data
# divide the data into attributes and labels
X = dataset.drop('Petrol_Consumption', axis=1)  
y = dataset['Petrol_Consumption']  

# dividing data into training and testing set
from sklearn.model_selection import train_test_split  
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, 
                                                    random_state=0)  

# Training and Making Predictions
# Note : we will using DecisionTreeRegressor class, not DecisionTreeClassifier



# To make predictions on the test set, 


# Now let's compare some of our predicted values with the actual values 



**Note** : 

that in your case the records compared may be different, depending upon the training and testing split. Since the train_test_split method randomly splits the data we likely won't have the same training and test sets. For train_test_split with random_state=0 , you would get the same results.


In [ ]:
# Evaluating the Algorithm
from sklearn import metrics  
print('Mean Absolute Error:', metrics.mean_absolute_error(y_test, y_pred))  
print('Mean Squared Error:', metrics.mean_squared_error(y_test, y_pred))  
print('Root Mean Squared Error:', np.sqrt(metrics.mean_squared_error(y_test, y_pred)))  

In [8]:
# doing minimum necessary imports

import pandas as pd                            # for loading and analysing data
import matplotlib.pyplot as plt                # for data visualization 
from sklearn.tree import DecisionTreeRegressor # Our Decision Tree classifier

In [9]:
!pip install Quandl ##Use Colab

Defaulting to user installation because normal site-packages is not writeable
You should consider upgrading via the '/usr/local/bin/python3 -m pip install --upgrade pip' command.


In [10]:
import quandl        # Stock market API for fetching Data

In [12]:
# ### import Quandl into your program as
import quandl  # Stock market API for fetching Data

# you can then fetch the stock data directly into your code as :
quandl.ApiConfig.api_key = 'cKoVfJoyzLxzqzsgb1Uz'   ## enter your key 
#stock_data = quandl.get("BSE/HDFC-BANK-LTD", start_date="2020-07-16", end_date="2021-07-16")  
stock_data = quandl.get('BSE/BOM500180', start_date='2020-10-13', end_date='2021-10-13')
# choose upto yesterday's date 
print(stock_data.head())   ### Let's see the data
print("----------------------")
print(stock_data.shape)   ### Let's see the data

               Open     High      Low    Close      WAP  No. of Shares  \
Date                                                                     
2020-10-13  1214.15  1221.50  1195.10  1198.15  1206.64       201283.0   
2020-10-14  1193.30  1214.80  1175.85  1210.85  1193.66       246863.0   
2020-10-15  1214.80  1216.45  1164.40  1169.15  1190.37       382299.0   
2020-10-16  1182.00  1203.00  1172.95  1199.00  1192.02       401346.0   
2020-10-19  1220.00  1228.80  1193.00  1203.35  1206.85       820336.0   

            No. of Trades  Total Turnover  Deliverable Quantity  \
Date                                                              
2020-10-13         6118.0     242876052.0               43263.0   
2020-10-14         8946.0     294671648.0               51586.0   
2020-10-15        11771.0     455077129.0              154798.0   
2020-10-16        12293.0     478414159.0              212464.0   
2020-10-19        24011.0     990019613.0              140712.0   

           

In [13]:
## working with shift()  --> https://www.geeksforgeeks.org/python-pandas-dataframe-shift/
print(stock_data[['Open', 'Close']].tail(4))
shifted_Open_Close = stock_data.loc[:,['Open', 'Close']].shift(-1)
print(shifted_Open_Close.tail(4))
print("-------------------------------")

shifted_Stock_Data = stock_data.copy()

print("-------------------------------")
shifted_Stock_Data['Open'] = shifted_Open_Close['Open']
print(shifted_Stock_Data[['Open', 'Close']].tail(4))

              Open    Close
Date                       
2021-10-08  1612.0  1602.20
2021-10-11  1603.0  1634.35
2021-10-12  1628.0  1629.30
2021-10-13  1637.0  1639.05
              Open    Close
Date                       
2021-10-08  1603.0  1634.35
2021-10-11  1628.0  1629.30
2021-10-12  1637.0  1639.05
2021-10-13     NaN      NaN
-------------------------------
-------------------------------
              Open    Close
Date                       
2021-10-08  1603.0  1602.20
2021-10-11  1628.0  1634.35
2021-10-12  1637.0  1629.30
2021-10-13     NaN  1639.05


In [14]:
## would be used later for testing purpose 
## last date value
print(shifted_Stock_Data.values[-1:])

[[           nan 1.64800000e+03 1.63000000e+03 1.63905000e+03
  1.63889000e+03 8.65720000e+04 4.11000000e+03 1.41881716e+08
  3.48300000e+04 4.02300000e+01 1.80000000e+01 2.05000000e+00]]


In [15]:
# How many samples do we have ?
#stock_data.shape
shifted_Stock_Data.shape

(250, 12)

In [16]:
# checking whether any column or feature has missing values
#stock_data.isnull().sum()

shifted_Stock_Data.isnull().sum()

Open                         1
High                         0
Low                          0
Close                        0
WAP                          0
No. of Shares                0
No. of Trades                0
Total Turnover               0
Deliverable Quantity         0
% Deli. Qty to Traded Qty    0
Spread H-L                   0
Spread C-O                   0
dtype: int64

In [17]:
shifted_Stock_Data.dropna(inplace=True)
shifted_Stock_Data.isnull().sum()

Open                         0
High                         0
Low                          0
Close                        0
WAP                          0
No. of Shares                0
No. of Trades                0
Total Turnover               0
Deliverable Quantity         0
% Deli. Qty to Traded Qty    0
Spread H-L                   0
Spread C-O                   0
dtype: int64

In [18]:
# Let's select our features

X = shifted_Stock_Data.drop(['Open'] , axis=1) 
y = shifted_Stock_Data.loc[ : ,'Open']

In [19]:
X.head(2) # head() shows the earliest 2 records

High      Low    Close      WAP  No. of Shares  No. of Trades  \
Date                                                                          
2020-10-13  1221.5  1195.10  1198.15  1206.64       201283.0         6118.0   
2020-10-14  1214.8  1175.85  1210.85  1193.66       246863.0         8946.0   

            Total Turnover  Deliverable Quantity  % Deli. Qty to Traded Qty  \
Date                                                                          
2020-10-13     242876052.0               43263.0                      21.49   
2020-10-14     294671648.0               51586.0                      20.90   

            Spread H-L  Spread C-O  
Date                                
2020-10-13       26.40      -16.00  
2020-10-14       38.95       17.55

In [20]:
y.head(2) # latest 2 stock prices

Date
2020-10-13    1193.3
2020-10-14    1214.8
Name: Open, dtype: float64

In [21]:
y.tail(2) # latest 2 stock prices

Date
2021-10-11    1628.0
2021-10-12    1637.0
Name: Open, dtype: float64

In [22]:
# split the entire data into Training and Test 
# keep 80% for training and 20% for Testing
# so the test_size = 0.2
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test = train_test_split(X, y, 
                                                 test_size = 0.2, 
                                                 random_state = 0)

In [23]:
# Let's fit our DecisionTree Model over the training data.
regressor = DecisionTreeRegressor() # making the object of DecisionTreeRegressor
regressor.fit(x_train,y_train)

DecisionTreeRegressor()

In [24]:
# Get the predictions on the test set 
y_pred = regressor.predict(x_test)

# Evaluating the Algorithm

from sklearn import metrics  
import numpy as np
print('Root Mean Squared Error:', np.sqrt(metrics.mean_squared_error(y_test, y_pred))) 

Root Mean Squared Error: 16.271318323971176


In [25]:
# Now , whats the mean of 'Close_Price' ??
# please note we will find the mean of the Close_Price of the dataset.
# Spliting data into Training and Testing is only for the purpose evaluating the Model
stock_data['Close'].mean()

1466.294199999999

<font color='green'> <b>Analysis : </b> Now the 10% of ---- is ~ <b>-----</b>. Our RMSE was ~<b>-----</b>. <br>I am very happy that its within 10% range. That means our <u>model</u> is doing a good job. 

In [27]:
# Trying to predict Tommorow's rate, i.e 14th Oct 2021 in my case. 
#test_data = [[1530.0,	1519.35,	1521.7,	1524.06,	91268.0,	4591.0,	139097555.0]]
#test_data = [[1522.0,	1530.0,	1519.35,	1524.06,	91268.0,	4591.0,	139097555.0,	32924.0,	36.07,	10.65,	-0.3]]

## testing for shifted data
test_data = [[ 1.64485000e+03, 1.60000000e+03, 1.63435000e+03,  1.63190000e+03, 1.44074000e+05, 6.35500000e+03, 
              2.35114992e+08, 7.79400000e+04, 5.41000000e+01, 4.48500000e+01, 3.13500000e+01]]

regressor.predict(test_data)

/usr/local/lib/python3.7/site-packages/sklearn/base.py:451: UserWarning: X does not have valid feature names, but DecisionTreeRegressor was fitted with feature names
  "X does not have valid feature names, but"


array([1628.])